xds/src/test/java/io/grpc/xds/CdsLoadBalancerTest.java

```diff
@@ -46,547 +30,498 @@
+                  .setEdsServiceName(edsServiceName)
+                  .setLbPolicy("round_robin")  // only supported policy
+                  .setLrsServerName(lrsServerName)
+                  .setUpstreamTlsContext(tlsContext)
+                  .build());
+        }
+      });
+    }
+
+    void deliverResourceNotFound() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onResourceDoesNotExist(CLUSTER);
+        }
+      });
+    }
+
+    void deliverError(final Status error) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onError(error);
+        }
+      });
+    }
+  }
+
+  private final class FakeLoadBalancerProvider extends LoadBalancerProvider {
+    private final String policyName;
+
+    FakeLoadBalancerProvider(String policyName) {
+      this.policyName = policyName;
     }
 
     @Override
     public LoadBalancer newLoadBalancer(Helper helper) {
-      edsLbHelpers.add(helper);
-      LoadBalancer edsLoadBalancer = mock(LoadBalancer.class);
-      edsLoadBalancers.add(edsLoadBalancer);
-      return edsLoadBalancer;
+      FakeLoadBalancer balancer = new FakeLoadBalancer(policyName, helper);
+      childBalancers.add(balancer);
+      return balancer;
     }
-  };
 
-  private final LoadBalancerProvider fakeRoundRobinLbProvider = new LoadBalancerProvider() {
     @Override
     public boolean isAvailable() {
       return true;
     }
 
     @Override
     public int getPriority() {
-      return 5;
+      return 0;  // doesn't matter
     }
 
     @Override
     public String getPolicyName() {
-      return "round_robin";
+      return policyName;
+    }
+  }
+
+  private final class FakeLoadBalancer extends LoadBalancer {
+    private final String name;
+    private final Helper helper;
+    private Object config;
+    private Status upstreamError;
+    private boolean shutdown;
+
+    FakeLoadBalancer(String name, Helper helper) {
+      this.name = name;
+      this.helper = helper;
     }
 
     @Override
-    public LoadBalancer newLoadBalancer(Helper helper) {
-      return mock(LoadBalancer.class);
+    public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+      config = resolvedAddresses.getLoadBalancingPolicyConfig();
     }
 
     @Override
-    public ConfigOrError parseLoadBalancingPolicyConfig(
-        Map<String, ?> rawLoadBalancingPolicyConfig) {
-      return ConfigOrError.fromConfig("fake round robin config");
+    public void handleNameResolutionError(Status error) {
+      upstreamError = error;
     }
-  };
 
-  private final SynchronizationContext syncContext = new SynchronizationContext(
-      new Thread.UncaughtExceptionHandler() {
+    @Override
+    public void shutdown() {
+      shutdown = true;
+      childBalancers.remove(this);
+    }
+
+    void deliverSubchannelState(final Subchannel subchannel, ConnectivityState state) {
```

Comment:
Normally if delivering a picker with a subchannel, the state is READY. If delivering IDLE or TRANSIENT_FAILURE, the picker is buffer picker or error picker.

Reply:
That's fine, it doesn't violate the API spec. Also, there is only one usage here, which is delivering a READY state with a subchannel.